File: 216.combination-sum-iii.java
/*
 * @lc app=leetcode id=216 lang=java
 *
 * [216] Combination Sum III
 */

// @lc code=start
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    LinkedList<Integer> temp = new LinkedList<>();
    
    public List<List<Integer>> combinationSum3(int k, int n) {
        dfs(1, k, n);
        return ans;
    }
    
    void dfs(int start, int k, int n) {
        if(n<0) return;
        if(k==0) {
            if(n==0) ans.add( new ArrayList(temp) );
            return;
        }
        for(int i=start; i<=9; ++i) {
            temp.add(i);
            dfs(i+1, k-1, n-i);
            temp.removeLast();
        }
    }
}
// @lc code=end

File: solutions/216.combination-sum-iii.rb
# @param {Integer} k
# @param {Integer} n
# @return {Integer[][]}
def combination_sum3(k, n)
  ans = []
  temp = []
  dfs = lambda do |start, cnt, target|
    return if target < 0

    if cnt.zero?
      ans << temp.dup if target.zero?
      return
    end

    (start..9).each do |i|
      temp << i
      dfs.call(i+1, cnt-1, target-i)
      temp.pop
    end
  end

  dfs.call(1, k, n)
  ans
end
